File: tradesmanager/ios/TradesManager/Shared/LocalizedText.swift
import Foundation

/// Text carried in every language it has been written in, keyed by language code.
///
/// The Kotlin side stores exactly this shape (see `LocalizedText.kt`), and both
/// read the same JSON, so a change to the resolution rule has to be made in two
/// places on purpose rather than in one by accident. The rules are:
/// exact match, then the base language of a regional tag, then the source
/// language, then any translation present.
public typealias LocalizedText = [String: String]

public enum Localised {

    /// The language the catalogues are authored in.
    public static let sourceLanguage = "en"

    public static func resolve(_ text: LocalizedText, languageCode: String) -> String {
        if let exact = text[languageCode], !exact.isEmpty { return exact }

        let base = String(languageCode.prefix(while: { $0 != "-" }))
        if let baseMatch = text[base], !baseMatch.isEmpty { return baseMatch }

        if let source = text[sourceLanguage], !source.isEmpty { return source }

        return text.values.first(where: { !$0.isEmpty }) ?? ""
    }

    /// Every translation joined and lowercased — the search index, so a search
    /// typed in Hebrew still finds an item the user remembers in English.
    public static func searchable(_ text: LocalizedText) -> String {
        text.values.map { $0.lowercased() }.joined(separator: " ")
    }
}

public extension Dictionary where Key == String, Value == String {
    /// `item.names.resolved(in: "he")`
    func resolved(in languageCode: String) -> String {
        Localised.resolve(self, languageCode: languageCode)
    }
}

File: tradesmanager/android/app/src/main/java/il/co/tradesmanager/core/i18n/LocalizedText.kt
package il.co.tradesmanager.core.i18n

/**
 * Text carried in every language it has been written in, keyed by language tag.
 *
 * Catalogue content is localised as *data*, not as Android resources: the
 * catalogues ship as JSON that iOS bundles unchanged, and a user typing their
 * own item name in Arabic must be able to store it next to a Hebrew one. A
 * fourth language therefore needs no schema change — only more keys.
 */
typealias LocalizedText = Map<String, String>

/** The source language, used as the fallback before giving up on order. */
const val SOURCE_LANGUAGE: String = "en"

/**
 * Best available rendering of this text for [languageTag].
 *
 * Falls back to the base language of a regional tag ("he-IL" -> "he"), then to
 * the source language, then to any translation present, so a partially
 * translated catalogue still shows something rather than an empty row.
 */
fun LocalizedText.resolve(languageTag: String): String {
    this[languageTag]?.takeIf { it.isNotBlank() }?.let { return it }
    val base = languageTag.substringBefore('-')
    this[base]?.takeIf { it.isNotBlank() }?.let { return it }
    this[SOURCE_LANGUAGE]?.takeIf { it.isNotBlank() }?.let { return it }
    return values.firstOrNull { it.isNotBlank() }.orEmpty()
}

/** Every translation joined, lowercased — what the search index is built from. */
fun LocalizedText.searchable(): String =
    values.joinToString(" ") { it.lowercase() }

fun localizedTextOf(vararg pairs: Pair<String, String>): LocalizedText =
    pairs.filter { it.second.isNotBlank() }.toMap()
